### ISSUE TITLE
Add VkPhysicalDevicePortabilitySubsetFeaturesKHR:: workgroupMatrices capability

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/Vulkan-Portability#33]
Based on CTS testing, and [this SPIRV-Cross discussion](https://github.com/KhronosGroup/SPIRV-Cross/issues/1770), it appears that Metal currently has a Catch-22 oversight, where initialization of locally declared `threadgroup` variables is required (like C++ statics), but Metal matrix variables do not have the necessary constructors in the `threadgroup` address space.

As discussed in the SPIRV-Cross thread, although it might be possible to create a heroic workaround in SPIRV-Cross, it is not clear that it would have real-world requirements for going that far.

An alternative is to add a `VkPhysicalDevicePortabilitySubsetFeaturesKHR:: workgroupMatrices` capability, where an implementation can indicate that SPIR-V can or cannot include matrices in a Workgroup variable.

**Comment by kvark** [comment]
Something doesn't seem right here. I can use matrices in `theradgroup` memory in Metal, e.g.
```cpp
#include <metal_stdlib>
#include <simd/simd.h>

struct Foo {
    metal::float2x2 m;
};

kernel void main1(
  threadgroup Foo const& foo
) {
    float x;
    float _e6 = foo.m[0].x;
    x = _e6;
    return;
}
```

**Comment by cdavis5e** [comment]
But you can't do this:

```c++
#include <metal_stdlib>
#include <simd/simd.h>

struct Foo {
    metal::float2x2 m;
};

kernel void main1() {
    threadgroup Foo foo;
    float x;
    float _e6 = foo.m[0].x;
    x = _e6;
    return;
}
```

In practice, we're more likely to see this latter pattern, because AFAIK SPIR-V has no equivalent for passing the TGSM block as an argument like Metal.

**Comment by cdavis5e** [comment]
Another thing about that TGSM parameter is that you have to set it up from the `MTLComputeCommandEncoder` with `-[MTLComputeCommandEncoder setThreadgroupMemoryLength:atIndex:]`. Again, Vulkan has no equivalent.

**Comment by billhollings** [comment]
> But you can't do this:
> 
> ```c++
>     threadgroup Foo foo;
> ```

And specifically, the CTS tests under review set values within the `threadgroup` matrix:

```c++
     threadgroup Foo foo;
     foo.m[0].x = 1.0;
```

**Comment by kvark** [comment]
What is GLSL/SPIR-V code used by these tests? I'm not getting why Metal code would need to declare "threadgroup Foo foo;" as a local variable (since it's not local, it's group-shared).

**Comment by billhollings** [comment]
>What is GLSL/SPIR-V code used by these tests? I'm not getting why Metal code would need to declare "threadgroup Foo foo;" as a local variable (since it's not local, it's group-shared).

The catalyst for this (including relevant SPIR-V) is in the [back-link mention](https://github.com/KhronosGroup/SPIRV-Cross/issues/1770) I made above (just above your question). That back-link is actually there to [reference](https://github.com/KhronosGroup/SPIRV-Cross/issues/1770#issuecomment-952011102) the thought you made above as leading to a potential solution.

Happy to have any suggestions about how that can be dealt with.

**Comment by kvark** [comment]
I'm just trying to understand the problem. It would help to first provide a piece of GLSL code that, when translated to SPIR-V and then to Metal, causes a problem. I tried converting the [SPIR-V to GLSL](https://shader-playground.timjones.io/1f0bb57cd8774d76930db79367a83e71) and saw this:
```
struct _2
{
    vec4 _m0;
    mat3x2 _m1;
    bvec4 _m2;
};
shared _2 _154;
```

So there is a global variable with `shared` qualifier, and it contains a matrix. Metal doesn't have global variables, so this one is expected to go the same route as all the resources (e.g. uniform buffers) - via function parameters. Doing this doesn't require any default constructors.

So where does the problem come from?

**Comment by billhollings** [comment]
>So where does the problem come from?

SPIRV-Cross converts the SPIR-V to include the following GLSL:

```
struct _152
{
    mediump vec4 _m0;
    mat3x2 _m1;
    bvec4 _m2;
};

shared _152 _154;

void main()
{
    _154._m0 = vec4(1.0, -5.0, -9.0, -5.0);
    _154._m1 = mat3x2(vec2(1.0, -7.0), vec2(1.0, 2.0), vec2(8.0, 7.0));
    _154._m2 = bvec4(false, true, false, false);
    ...
    vec4 _184 = _154._m0;
```

and the corresponding converted MSL includes:

```
struct _152
{
    float4 _m0;
    float3x2 _m1;
    bool4 _m2;
};

kernel void main0(device _210& _212 [[buffer(1)]])
{
    threadgroup _152 _154;
    _154._m0 = float4(1.0, -5.0, -9.0, -5.0);
    _154._m1 = float3x2(float2(1.0, -7.0), float2(1.0, 2.0), float2(8.0, 7.0));
    _154._m2 = short4(bool4(false, true, false, false));
    ...
    float4 _184 = _154._m0;
```

The MSL error results from the declaration of `_154`. It requires a default constructor, because MSL `threadgroup` local variables are initialized. Unfortunately, the default constructor of `_152` is deleted because `float3x2` does not have a default constructor.

Your focus on the `threadgroup` variable being in a buffer does offer one possible way through this, but it would require elevating the local variable to an input/output variable, and coordinating through the app to provide that buffer.

**Comment by kvark** [comment]
Yeah, I'm totally missing why SPIRV-Cross tries to make it a local variable. All global declarations in SPIR-V pretty much need to go as arguments for MSL, I don't understand why `threadgroup` should be an exception.

**Comment by HansKristian-Work** [comment]
> I don't understand why `threadgroup` should be an exception

Because threadgroup is not part of the interface with host. Similar to how we declare Private variables. You don't pass Private variables by main parameter. Only global variables which are IO or descriptors are main() parameters. Sure, Metal has this weird threadgroup-by-pointer-allocated-from-host thing, but that's not mandatory and kinda awkward. AFAIK, it's a hack to workaround the lack of spec constant array size of threadgroup storage when making the workgroup size a spec constant.

**Comment by kvark** [comment]
Not sure where "threadgroup-by-pointer-allocated-from-host" comes from. Are you talking about [setThreadgroupMemoryLength(_:index:)](https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443142-setthreadgroupmemorylength)? There is nothing about the host allocation here that I can see.

Can MoltenVK use this for all threadgroup memory? Would there be any downside for doing so?

**Comment by cdavis5e** [comment]
I'm not certain this is needed anymore. I actually had to implement a workaround to solve `shaderSharedCompositeVariables`--and then Apple finally did what I asked them to and added default matrix constructors in the `threadgroup` AS.

**Comment by billhollings** [comment]
> Apple finally did what I asked them to

Would we still need the flag for older OS's then?

**Comment by cdavis5e** [comment]
> Would we still need the flag for older OS's then?

From the prior message:

> I actually had to implement a workaround to solve `shaderSharedCompositeVariables`

So probably not.
